openfl/interface/aggregation_functions/fed_bagging.py

```diff
@@ -0,0 +1,130 @@
+                    Examples for `torch.nn.Module`s: 'conv1.weight','fc2.bias'.
+                - 'round': 0-based number of round corresponding to this
+                    tensor.
+                - 'tags': tuple of tensor tags. Tags that can appear:
+                    - 'model' indicates that the tensor is a model parameter.
+                    - 'trained' indicates that tensor is a part of a training
+                        result.
+                        These tensors are passed to the aggregator node after
+                        local learning.
+                    - 'aggregated' indicates that tensor is a result of
+                        aggregation.
+                        These tensors are sent to collaborators for the next
+                        round.
+                    - 'delta' indicates that value is a difference between
+                        rounds for a specific tensor.
+                    also one of the tags is a collaborator name
+                    if it corresponds to a result of a local task.
+
+                - 'nparray': value of the tensor.
+            tensor_name: name of the tensor
+            fl_round: round number
+            tags: tuple of tags for this tensor
+        Returns:
+            bytearray: aggregated tensor
+        """
+
+        global_model = get_global_model(db_iterator, fl_round)
+
+        if (
+            isinstance(global_model, np.ndarray) and global_model.size == 0
+        ) or global_model is None:
+            for local_tensor in local_tensors:
+                local_tree_bytearray = bytearray(local_tensor.tensor.astype(np.uint8).tobytes())
+                local_tree_json = json.loads(local_tree_bytearray)
+
+                if (
+                    isinstance(global_model, np.ndarray) and global_model.size == 0
+                ) or global_model is None:
+                    # the first tree becomes the global model
+                    global_model = local_tree_json
+                else:
+                    # append subsequent trees to global model
+                    local_model = local_tree_json
+                    local_trees = local_model["learner"]["gradient_booster"]["model"]["trees"]
+                    global_model = append_trees(global_model, local_trees)
+        else:
+            global_model_bytearray = bytearray(global_model.astype(np.uint8).tobytes())
+            global_model = json.loads(global_model_bytearray)
+
+            for local_tensor in local_tensors:
+                local_tree_bytearray = bytearray(local_tensor.tensor.astype(np.uint8).tobytes())
+                local_trees = json.loads(local_tree_bytearray)
+                global_model = append_trees(global_model, local_trees)
+
+        global_model_json = json.dumps(global_model)
+        global_model_bytes = global_model_json.encode("utf-8")
+
+        return np.frombuffer(global_model_bytes, dtype=np.uint8).astype(np.float32)
```

Comment:
I have to admit I'm a bit lost here... What is of type `np.uint8`, and what is of type `np.float32`? Don't we risk loss of precision in the transformation?

Reply:
There is some inefficiency happening here that can probably be optimized. Before getting sent to `fed_bagging.py`, each collaborator trains their own decision tree. XGBoost models are accessed with `.save_raw()`, which returns a bytearray (is a sequence of bytes (`uint8`)). Almost all of OpenFL's pipelines will cast any data getting sent across the federation to `np.float32` before converting to bytes (ex. [pipeline.py#L72](https://github.com/securefederatedai/openfl/blob/develop/openfl/pipelines/pipeline.py#L72)). In theory, the XGBoost model bytearray can just be converted to bytes and sent directly, but this would require changes in the pipeline to relax their conditionals or else it'll through an error since it's expecting `np.float32`. In addition, models are stored as np arrays in the TensorDB and the aggregated model/values get converted to an np array (see: [tensor_db.py#L247](https://github.com/securefederatedai/openfl/blob/develop/openfl/databases/tensor_db.py#L247)). So a lot of these initial steps were in an effort to preserve the primary infrastructure (tensordb and pipelines). To answer your question though, the bytearray cannot be directly converted to `np.float32` and therefore is converted to np.uint8 before being casted as `np.float32`. There is a risk of precision loss, but this would happen during the pipeline anyway, so doing it in the runner allows us to check for precision loss and raise a warning (I added a `conversion_check` function to do this at both the runner and the aggregation function).

In the future, I think we should aim to relax and/or otherwise address the conversion of data throughout the entire OpenFL work flow. I'll admit I need to dive a bit deeper into this myself since XGBoost was the first instance of having to deal with these types of conversions.

Reply:
Thanks for the detailed explanation, @kta-intel , this helps!